src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex.ExtensionMethodInfo.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.FindSymbols
+{
+    internal partial class SyntaxTreeIndex
+    {
+        private readonly struct ExtensionMethodInfo
+        {
+            // We devide extension methods into two categories, simple and complex, for filtering purpose.
+            // Whether a method is simple is determined based on if we can determine it's target type easily
+            // with a pure text matching. For complex methods, we will need to rely on symbol to decide if it's 
+            // feasible.
+            //
+            // Complex methods include:
+            // - Method declared in the document which includes using alias directive
+            // - Generic method
```

Comment:
generic method are fine.  it's only bad if hte if the first param is a TypeParamete right?

Reply:
Correct. I make it complex (and same for using alias) just so I can establish a perf baseline first. If those turns out to be very rare and doesn't really cause much problems in Roslyn.sln, then we might be able to delay the fix until other bigger issues got addressed.

Reply:
i guess my point is this:  generic methods aren't actually a problem. it's only about if hte *first parameter* is a 'named type' or not. if it's a named-type, it's simple, if it's not, then it's complex.

Reply:
Totally agree. What I meant is I plan to make improvement like this to the filtering, once I get some idea on the perf of current implementation :)

Reply:
k.  can you remove this now? :)

Reply:
Done